src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -612,6 +663,166 @@ private BoundExpression CreateAndPopulateArray(BoundCollectionExpression node, A
                 arrayType);
         }
 
+        /// <summary>
+        /// Returns true if type is convertible to Span or ReadOnlySpan.
+        /// If non-identity conversion, also returns a non-null asSpanMethod.
+        /// </summary>
+        private bool TryGetSpanConversion(TypeSymbol type, out MethodSymbol? asSpanMethod)
+        {
+            if (type is ArrayTypeSymbol { IsSZArray: true } arrayType
+                && _factory.WellKnownMethod(WellKnownMember.System_Span_T__ctor_Array, isOptional: true) is { } spanCtorArray)
+            {
+                asSpanMethod = spanCtorArray.AsMember(spanCtorArray.ContainingType.Construct(arrayType.ElementType));
```

Comment:
How do we know the element type is always going to be a valid type argument? Can we assert something?

Reply:
I would be fine with asserting, for example, that element type is implicitly convertible to object.

Reply:
This was a faulty assumption. The conversion to object is now checked in retail builds and if it isn't what we expect then we don't use either the ToArray or CopyTo optimization paths.

Reply:
Do we have a test that hits this? (It looks like the existing tests were unchanged, other than setting the target framework.)

Reply:
The new tests require the object conversion check in order to work properly.